### PR TITLE
Dev/rgb matrix overhaul

### PR DESCRIPTION
**Pull request by XScorpion2**
## Description

Large Overhaul of the rgb_matrix code to be realtime and performant. Builds on top of https://github.com/qmk/qmk_firmware/pull/5304

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: common_features.mk [hunk 1]
```diff
@@ -144,6 +144,9 @@ endif
     SRC += $(QUANTUM_DIR)/rgb_matrix.c
     SRC += $(QUANTUM_DIR)/rgb_matrix_drivers.c
     CIE1931_CURVE = yes
+    ifeq ($(strip $(RGB_MATRIX_KEYPRESSES)), yes)
```
Comment (drashna) [inline common_features.mk]: Just curious, why is this added/moved to here? 
Usually, we have the config.h file for settings like this.

Comment (XScorpion2) [reply common_features.mk]: I can remove it, added it as a hack tbh

Comment (drashna) [reply common_features.mk]: Yeah, I think it would be for the best to remove it

**Comment by XScorpion2** [comment]
Ready for review as this is about as performant as I can get it without getting to crazy. This should be merged after https://github.com/qmk/qmk_firmware/pull/5328 and https://github.com/qmk/qmk_firmware/pull/5330 so I can fix any merge errors that will occur from this pr when they land. I've already manually merged those into another test branch and verified they will work with my changes here with a few minor adjustments.

**Comment by XScorpion2** [comment]
https://github.com/qmk/qmk_firmware/pull/5330 was merged to master, master has been merged into here and the new matrix anim effect has been updated to match this refactor work.

File: quantum/rgb_matrix_animations/solid_reactive_simple_anim.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+#ifdef RGB_MATRIX_KEYPRESSES
+#ifndef DISABLE_RGB_MATRIX_SOLID_REACTIVE_SIMPLE
+
+extern rgb_config_t rgb_matrix_config;
+extern last_hit_t g_last_hit_tracker;
+
+bool rgb_matrix_solid_reactive(bool init, uint8_t iter) {
```
Comment (drashna) [inline quantum/rgb_matrix_animations/solid_reactive_simple_anim.h]:
```suggestion
bool rgb_matrix_solid_reactive_simple(bool init, uint8_t iter) {
```

Comment (XScorpion2) [reply quantum/rgb_matrix_animations/solid_reactive_simple_anim.h]: hmm, not sure how I screwed that up, but thanks for catching it.

**Comment by drashna** [comment]
At this point, I'm not sure if anything else needs to be changed/added/improved here. At least code wise.

However, could you update the documentation for the RGB Matrix? 

There is at least one new setting that has been added, and the effects order has been changed. 
If you could update the docs to reflect these changes and additions, that would be awesome!

**Comment by XScorpion2** [comment]
Yup, will he working on docs over the next few days or so

**Comment by patrickmt** [comment]
I benchmarked all these animations on a CTRL keyboard using @XScorpion2's dev-rgb_ctrl branch and RGB_MATRIX_LED_PROCESS_LIMIT set to 15 to mimic the Massdrop lighting configuration. The CTRL has 119 RGB LEDs and runs at 40MHz. This means that after 8 calls of rgb_matrix_task, all LEDs are processed and ready to output.

Most effects work very well, taking around 1.9ms to 2.6ms for animations to complete. Note that these times include any other code running between effect processing, including hardware ADC sampling tasks. Actual effect only processing times are more around 1.2ms to 1.9ms. The few issues I saw were with digital rain and multisplash effects where those took around 6ms. These tests were performed on the branch at 53053ec.

Because of various time sensitive tasks taking place on the CTRL, breaking up the LED processing has helped tremendously in making it operate as intended while using the RGB Matrix effects. One thing missing still is the ability to control the rate of flushes instead of flushing as fast as possible. This would allow one to target an update rate while giving the system time to operate without possibly expensive LED processing interfering.

I've discussed all this with @XScorpion2 and he said he will add in the flush rate timing ability.

File: quantum/rgb_matrix.c [hunk 3]
```diff
@@ -250,12 +250,12 @@ static void rgb_task_timers(void) {
...
     last_hit_buffer.tick[i] += deltaTime;
   }
-#endif
+#endif // defined(RGB_MATRIX_KEYPRESSES) || defined(RGB_MATRIX_KEYRELEASES)
 }
 
 static void rgb_task_sync(void) {
   // next task
-  if (timer_elapsed32(g_rgb_counters.tick) > 16)
+  if (timer_elapsed32(g_rgb_counters.tick) > RGB_MATRIX_LED_FLUSH_LIMIT)
```
Comment (patrickmt) [inline quantum/rgb_matrix.c]: Should be >= RGB_MATRIX_LED_FLUSH_LIMIT. Verified using scope.

Comment (XScorpion2) [reply quantum/rgb_matrix.c]: good point

Comment (XScorpion2) [reply quantum/rgb_matrix.c]: fixed

**Comment by XScorpion2** [comment]
Still want to document each of the animations, but other than that I think it is good to go.

File: quantum/rgb_matrix_animations/breathing_anim.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+extern rgb_config_t rgb_matrix_config;
+
+bool rgb_matrix_breathing(bool init, uint8_t iter) {
+  RGB_MATRIX_USE_LIMITS(led_min, led_max);
+
+  uint16_t time = scale16by8(g_rgb_counters.tick, rgb_matrix_config.speed / 8);
+  uint8_t val = scale8(abs8(sin8(time) - 128) * 2, rgb_matrix_config.val);
+  HSV hsv = { rgb_matrix_config.hue, rgb_matrix_config.sat, val };
+  RGB rgb = hsv_to_rgb(hsv);
+  for (uint8_t i = led_min; i < led_max; i++) {
```
Comment (patrickmt) [inline quantum/rgb_matrix_animations/breathing_anim.h]: For some reason, each iteration of this loop is taking about 0.6ms, so total time taken is around `DRIVER_LED_TOTAL * 0.6ms`. I can see it as a choppy animation on the keyboard.

Comment (XScorpion2) [reply quantum/rgb_matrix_animations/breathing_anim.h]: `DRIVER_LED_TOTAL * 0.6ms` or `RGB_MATRIX_LED_PROCESS_LIMIT * 0.6ms`?
Either way that is super odd. I'll do another pass on this animation tonight as I'd like to get it to pause at the high end vs continual fluctuations as it does now.

Comment (patrickmt) [reply quantum/rgb_matrix_animations/breathing_anim.h]: DRIVER_LED_TOTAL, as in 119 * 0.6ms = 71.4ms for a breathing animation on the CTRL

Comment (XScorpion2) [reply quantum/rgb_matrix_animations/breathing_anim.h]: ew, will fix it tonight. was coding blind last night without testing the results, my bad

Comment (yanfali) [reply quantum/rgb_matrix_animations/breathing_anim.h]: I think the goal would be to get this under 33ms and aim for a steady 30fps. The closer you can get to 16.6ms/60fps the better it should look.

Comment (patrickmt) [reply quantum/rgb_matrix_animations/breathing_anim.h]: This animation was running at 0.6ms for the entire update (119 LEDs processed) in my bench marking from a few commits back before it was updated to batch processing. But now I see that simply the call to rgb_matrix_set_color_all() is taking that much time and then happening 119 times. I briefly looked into the logic myself and the compiled assembly but couldn't see anything fishy with how the loop was being assembled. I didn't try to compile without optimizations though, which might be a useful test.

Comment (XScorpion2) [reply quantum/rgb_matrix_animations/breathing_anim.h]: ...I'm dumb, that should have been changed to rgb_matrix_set_color(i... instead of rgb_matrix_set_color_all

Comment (XScorpion2) [reply quantum/rgb_matrix_animations/breathing_anim.h]: there, fixed. Now I'm not doing 119 * (119 / 15) iterations....

**Comment by patrickmt** [comment]
This PR is working very nice now, tested on both CTRL and ALT keyboards!

**Comment by XScorpion2** [comment]
Force push rebase on latest master

**Comment by XScorpion2** [comment]
Minor api change for future compatibility of things like this:
https://photos.app.goo.gl/waCkRJwsqjEBiFCCA

**Comment by XScorpion2** [comment]
Fixed a compile error that would occur if you had just `RGB_MATRIX_KEYRELEASES` defined

**Comment by drashna** [comment]
I noticed some issues with resuming from sleep on my board. I'm not sure if it's related to these changes, though.  Will test and let you know.

**Comment by drashna** [comment]
Lockup was a red herring.  It looks like it maybe some other change, and not this PR

**Comment by XScorpion2** [comment]
@drashna tease

**Comment by drashna** [comment]
> @drashna tease

Maybe, but I'd rather say something and have testing prove me wrong, then not say anything and have a bug merged in

**Comment by drashna** [comment]
Thanks!
